hir-analysis/src/def_use.rs

```diff
@@ -0,0 +1,424 @@
+use cranelift_entity::SecondaryMap;
+use intrusive_collections::{intrusive_adapter, LinkedListLink};
+use midenc_hir::{Block, BranchInfo, DataFlowGraph, Inst, Instruction, Value, ValueData};
+
+use crate::DominatorTree;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum ValueDef {
+    /// This corresponds to the original definition of a value as an instruction result
+    Inst {
+        /// The result index
+        num: u16,
+        /// The defining instruction
+        inst: Inst,
+    },
+    /// This corresponds to the original definition of a value as a block parameter
+    Param {
+        /// The parameter index
+        num: u16,
+        /// The defining block
+        block: Block,
+    },
+    /// This represents a reload of a spilled value
+    Reload {
+        /// The reload instruction
+        inst: Inst,
+    },
+    /// This definition corresponds to an implicit phi node in the dataflow graph.
+    ///
+    /// Specifically, this is like `Param`, but for a block parameter that doesn't yet exist,
+    /// but _must_ be created in order for SSA form to be preserved.
+    Phi {
+        /// The block in which this phi must be materialized as a block parameter
+        block: Block,
+    },
+}
+
+/// The def-use graph provides us with three useful pieces of information:
+///
+/// * The set of users for any value in a function
+/// * Whether a value is ever used
+/// * The nearest dominating definition for a specific use of a value, taking into account spills
+///
+/// It is computed by visiting the reachable blocks of a function, and building a graph of all the
+/// value definitions and their uses.
+#[derive(Default)]
+pub struct DefUseGraph {
```

Comment:
Currently, `DefUseGraph` is not being used, although I originally had intended to use it in the implementation of `RewriteSpills`. However, in the case of the `RewriteSpills`, it turned out that I could write a specialized traversal that was much more efficient than using the `DefUseGraph` for the same job, so I ended up setting it aside. I've kept it around though, because in the future we may add more analyses or optimizations for which the `DefUseGraph` will be essential.

The thing that probably will help make it click: our IR is represented as a dataflow graph, i.e. we have some set of instructions, and for each instruction, we know what values it depends on, and what values it produces. It is very easy for us to answer questions relative to a given instruction. The problem is that our representation doesn't provide us that same level of information for individual values, i.e. what defined a value, is that value used, and who uses it. The `DefUseGraph` is basically a structure whose sole responsibility is providing answers to those questions. In practice, you might also use the `DefUseGraph` to visit all the users of a value, and rewrite those users to use a new value instead, things like that.

Hope that helps!

Reply:
Thank you! It helped a lot!